### PR TITLE
Support loading BERT/RoBERTa/NeoX tokenizers without HF transformers

### PR DESCRIPTION
**Pull request by danieldk**
## Description

This change adds the `FromHFHub` mixin for tokenizers that supports loading of tokenizers without the `transformers` package installed. This works for every tokenizer type that we support, except for sentencepiece tokenizers, since we only have the upstream sentencepiece implementation that only loads from sentencepiece model files.

The `FromPretrainedHFTokenizers` implementations for these tokenizers are also rewritten, to use `FromHFHub` after serializing the tokenizer.

Smaller related changes:

- Also deserialize special pieces.
- Remove `slow` mark from tokenizer tests.

### Types of change

Enhancement

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: curated_transformers/tokenization/bert_tokenizer.py [hunk 1]
```diff
@@ -236,34 +248,48 @@ def from_files(
...
-        vocab = [None] * tokenizer.vocab_size  # type: ignore
-        for piece, idx in tokenizer.vocab.items():  # type: ignore
-            vocab[idx] = piece
-
-        bos_piece = tokenizer.cls_token  # type: ignore
-        eos_piece = tokenizer.sep_token  # type: ignore
-        unk_piece = tokenizer.unk_token  # type: ignore
-        lowercase = tokenizer.do_lower_case  # type: ignore
-        strip_accents = tokenizer.backend_tokenizer.normalizer.strip_accents  # type: ignore
+    def _convert_hf_tokenizer_json(cls: Type[Self], *, hf_tokenizer: Any) -> Self:
```
Comment (shadeMe) [inline curated_transformers/tokenization/bert_tokenizer.py:251]:
```suggestion
    def _convert_hf_tokenizer_json(cls: Type[Self], *, hf_tokenizer: Dict[str, Any]) -> Self:
```

File: curated_transformers/tokenization/gpt_neox_tokenizer.py [hunk 2]
```diff
@@ -1,49 +1,35 @@
...
-        """Construct a tokenizer from the vocabulary and merges files.
-
-        vocab_path (Path): path to the vocabulary file.
-        merges_path (Path): path to the merges file.
-        """
-        processor = ByteBPEProcessor.load_from_files(
-            vocab=vocab_path, merges=merges_path
-        )
-        return cls(processor=processor)
+    def _convert_hf_tokenizer_json(cls: Type[Self], *, hf_tokenizer: Any) -> Self:
```
Comment (shadeMe) [inline curated_transformers/tokenization/gpt_neox_tokenizer.py:18]:
```suggestion
    def _convert_hf_tokenizer_json(cls: Type[Self], *, hf_tokenizer: Dict[str, Any]) -> Self:
```

File: curated_transformers/tokenization/gpt_neox_tokenizer.py [hunk 3]
```diff
@@ -1,49 +1,35 @@
...
+        # TODO: Seems like we can't easily probe the model type?
```
Comment (shadeMe) [inline curated_transformers/tokenization/gpt_neox_tokenizer.py:19]: I [see](https://huggingface.co/EleutherAI/gpt-neox-20b/raw/main/tokenizer.json) a `type` entry in `hf_tokenizer["model"]`?

File: curated_transformers/tokenization/hf_hub.py [hunk 4]
```diff
@@ -1,11 +1,87 @@
...
+    """
+    Mixin class for downloading tokenizers from Hugging Face Hub.
+    """
+
+    @classmethod
+    def from_tokenizer_json_file(
+        cls: Type[SelfFromHFHub],
+        path: Path,
+    ) -> SelfFromHFHub:
+        """Construct a Hugging Face fast tokenizer JSON file.
```
Comment (shadeMe) [inline curated_transformers/tokenization/hf_hub.py:28]:
```suggestion
        """Construct a tokenizer from a Hugging Face fast tokenizer JSON file.
```

File: curated_transformers/tokenization/hf_hub.py [hunk 5]
```diff
@@ -1,11 +1,87 @@
...
+        """
+        with open(path, "r") as f:
+            tokenizer = json.load(f)
+
+        return cls._convert_hf_tokenizer_json(hf_tokenizer=tokenizer)
+
+    @classmethod
+    @abstractmethod
+    def _convert_hf_tokenizer_json(
+        cls: Type[SelfFromHFHub], *, hf_tokenizer: Any
```
Comment (shadeMe) [inline curated_transformers/tokenization/hf_hub.py:43]:
```suggestion
        cls: Type[SelfFromHFHub], *, hf_tokenizer: Dict[str, Any]
```

File: curated_transformers/tokenization/hf_hub.py [hunk 6]
```diff
@@ -1,11 +1,87 @@
...
+
+
+def _get_tokenizer_filepath(name: str, revision: str) -> str:
+    try:
+        return hf_hub_download(
+            repo_id=name, filename=HF_TOKENIZER_FILENAME, revision=revision
+        )
+    except:
+        raise ValueError(
+            f"Couldn't find a valid config for model `{name}` "
```
Comment (shadeMe) [inline curated_transformers/tokenization/hf_hub.py:82]:
```suggestion
            f"Couldn't find a valid tokenizer config for model `{name}` "
```

File: curated_transformers/tokenization/roberta_tokenizer.py [hunk 7]
```diff
@@ -83,26 +91,49 @@ def from_files(
             vocab=vocab_path, merges=merges_path
         )
         return cls(
-            processor=processor,
+            # This is a bit annoying, but we want to avoid these extremely
+            # overloaded constructor.
```
Comment (shadeMe) [inline curated_transformers/tokenization/roberta_tokenizer.py:95]:
```suggestion
            # overloaded constructors.
```

File: curated_transformers/tokenization/roberta_tokenizer.py [hunk 8]
```diff
@@ -83,26 +91,49 @@ def from_files(
...
     @classmethod
-    def convert_hf_tokenizer(cls: Type[Self], tokenizer: Any) -> Self:
-        serialized = tokenizer.backend_tokenizer.to_str(True)  # type: ignore
-        deserialized = json.loads(serialized)
-        vocab_merges = deserialized["model"]
-        merges = [tuple(merge.split(" ")) for merge in vocab_merges["merges"]]
-        processor = ByteBPEProcessor(vocab_merges["vocab"], merges)
-        bos_piece = tokenizer.bos_token  # type: ignore
-        eos_piece = tokenizer.eos_token  # type: ignore
+    def _convert_hf_tokenizer_json(cls: Type[Self], *, hf_tokenizer: Any) -> Self:
```
Comment (shadeMe) [inline curated_transformers/tokenization/roberta_tokenizer.py:103]:
```suggestion
    def _convert_hf_tokenizer_json(cls: Type[Self], *, hf_tokenizer: Dict[str, Any]) -> Self:
```

File: curated_transformers/tokenization/wordpiece_tokenizer.py [hunk 9]
```diff
@@ -12,13 +12,35 @@ class WordPieceTokenizer(Tokenizer):
...
+        # all tokens as a sanity check.
+        vocab_size = max(vocab.values()) + 1
+        pieces: List[Optional[str]] = [None] * vocab_size
+        for piece, idx in vocab.items():
+            pieces[idx] = piece
+
+        unused_indices = [str(id) for id, piece in enumerate(pieces) if piece is None]
+        if unused_indices:
+            raise ValueError(
+                f"Vocabulary contains unused indices: {', '.join(unused_indices)}"
```
Comment (shadeMe) [inline curated_transformers/tokenization/wordpiece_tokenizer.py:40]:
```suggestion
                f"WordPiece vocabulary contains unused indices: {', '.join(unused_indices)}"
```

File: curated_transformers/tokenization/gpt_neox_tokenizer.py (hunk 3 above)
Comment (danieldk) [reply curated_transformers/tokenization/gpt_neox_tokenizer.py:19]: But the type is `BPE`, which is not really specific. But I guess it's better than nothing, so I'll add that.

Comment (danieldk) [reply curated_transformers/tokenization/gpt_neox_tokenizer.py:19]: Though thinking more about this, the NeoX tokenizer is pretty much a plain Byte BPE tokenizer, so I'll move the loading methods there.

**Review by shadeMe — APPROVED** [review]

**Pull request merged**